### DISCUSSION TITLE
how do I merge non-connected manifolds

### DISCUSSION THREAD

**Discussion by savolla** [pyvista/pyvista#3308] [Q&A]
hey all! I have a file called `T_shape.stl` which consists of two non-connecting manifolds. As you can see in the image below the upper part and lower part are separate.  
![Screenshot from 2022-09-13 13-19-00](https://user-images.githubusercontent.com/22548773/189877853-7b8d0c0f-c859-4207-b22a-e84f6b901a25.jpg)
when I extract edges it gives me extra edges that I don't want. like this;

![Screenshot from 2022-09-13 13-28-19(1)](https://user-images.githubusercontent.com/22548773/189880102-f77c5cb5-de80-4456-bff5-b6c29264ce3b.jpg)
So I tryed to use `merge` method in pyvista but this method accepts two separate meshes to merge. I have a single mesh. how do I merge these manifolds? thanks!

**Comment by adeak** [comment]
Merging won't help, because it would give you exactly what you have now: two meshes with no actual interconnection in a single `PolyData`.

I don't know of a general solution, but my best guess would be to try [`boolean_union()`](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.boolean_union.html#pyvista.PolyDataFilters.boolean_union) on the two submeshes. And you can get the two submeshes with `mesh.split_bodies()` which should give you a 2-length `MultiBlock`. For best results the following should be true:
```py
multi = mesh.split_bodies()
assert len(multi) == 2
first = multi[0].triangulate()
second = multi[1].triangulate()
assert first.is_manifold
assert second.is_manifold
```
If the submeshes are not manifold you could still try calling `.clean()` on them: sometimes you have spurious point duplication that's taken care of by cleaning.

If the above all check out you might try the boolean operation. Still, it's not very stable to try to take the union of faces that overlap completely, in parallel (in fact VTK explicitly says that its algorithm is not prepared for this case). You could try shifting the vertical profile down a bit so that it has some overlap with the horizontal profile, and see what happens. In any case boolean operations are fickle so you can easily get weird results, a bunch of VTK errors, or even a segmentation fault.

---

I'll also note that for this specific mesh I'd start from scratch, creating the T profile by hand, and then extruding that in the perpendicular direction.

**Reply by savolla** [reply]
thanks for this extensive reply! `split_bodies()` method worked on my mesh. and it contained 2 separate bodies. The problem is that I cannot call `.is_manifold()` method either on `first` or `second` body because these bodies are `UnstructuredGrid` type. I also cannot call `first.boolean_union(second)` for the same reason. Is there a way to convert `UnstructuredGrid` to `PolyData`? I believe we are very close to the solution :D

**Reply by adeak** [reply]
Yeah, just call `.extract_geometry()` (I recall that working better in this use case than `.extract_surface()`, but there's both) on your unstructured grid, it should give you a `PolyData`.

**Reply by savolla** [reply]
running `.extract_geometry()` worked! I successfully merged those manifolds but now I have another issue. it is a little bit hard to see but there is one face missing. 

![ezgif com-gif-maker](https://user-images.githubusercontent.com/22548773/189920217-3c28a174-1d1a-4b7f-9690-10dd99d1df84.gif)

**Reply by adeak** [reply]
Hmm, it's indeed hard to see but I think I see what you mean.

Can you call `plot_normals(faces=True)` on the two submeshes and confirm that all the arrows point outwards? You might have to call `compute_normals(auto_orient_normals=True)` first to fix them.

**Reply by savolla** [reply]
I also tried to intersect my T shape with very huge Cube that swallows the T shape. the intersection should give me the T shape itself as a whole mesh but vtk cannot intersect them because there is no common intersecting points

**Reply by savolla** [reply]
all face normals are pointing to the correct direction. I think there is no easy way of doing this in pyvista. I don't even know the term for this operation. in CAD programs this is called *combining bodies*

here is face normals for `first`
![Screenshot from 2022-09-13 17-00-11](https://user-images.githubusercontent.com/22548773/189922672-1c94a20b-624b-45d8-ab17-85e91e8c0667.png)

and for `second`
![Screenshot from 2022-09-13 17-01-10](https://user-images.githubusercontent.com/22548773/189922698-50d3bb40-89f6-4293-b222-cbdf770f71bf.png)

**Reply by adeak** [reply]
OK, so if
1. both meshes are manifold
3. both meshes have normals pointing outward

then this might be a "weird result from boolean union", like I mentioned earlier. You should still try shifting the vertical mesh down a bit to get some finite intersection: the missing face might be the shortcoming of coincidental faces I also mentioned earlier. Normally I would expect the boolean union of two manifold surfaces to also be manifold, and your is clearly not.

But yeah, we might not have good tooling for this.

**Reply by savolla** [reply]
if I shift the one of the bodies I'll deform the part. this T shape is just a simple example. I have other more complex meshes in the same condition. Thank you so much for your support. Maybe I should fix this issue by using another tool. I'm closing this for now :)

**Reply by adeak** [reply]
Yeah, I only meant for debugging purposes, to figure out if it would work with a finite intersection.

I don't think you should mark this as accepted: others might have better ideas for how to fix this :)
